nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::SparseFormatError(SparseFormatErrorKind::IndexOutOfBounds));
+    /// ```
+    SparseFormatError(SparseFormatErrorKind),
+
+    /// Indicates that a wrong diagonal element has been provieded to the matrix
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_str;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// # use nalgebra::Complex;
+    /// let str = r#"
+    /// %%matrixmarket matrix coordinate real skew-symmetric
+    /// % skew-symmetric matrix can't have element on diagonal
+    /// 5 5 2
+    /// 1 1 10
+    /// 2 1 5
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<f64>(str);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::DiagonalError);
+    ///
+    /// let str = r#"
+    /// %%matrixmarket matrix coordinate complex hermitian
+    /// % hermitian matrix diagonal element must be a real number
+    /// 5 5 2
+    /// 1 1 10 2
+    /// 2 1 5 2
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<Complex<f64>>(str);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::DiagonalError);
+    /// ```
+    /// Here the skew matrix shouldn't have an element on the diagonal
+    DiagonalError,
+
+    /// Indicates [io error](`std::io::Error`), while reading the data from file.
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_file;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// let file_name = "whatever.mtx";
+    /// let matrix_error = load_coo_from_mm_file::<f64,_>(file_name);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::IOError(std::io::ErrorKind::NotFound));
+    /// ```
+    IOError(std::io::ErrorKind),
+
+    /// Indicates (skew-)symmetric (or hermitian) matrix is not lower triangle matrix.  
+    ///
+    /// Examples
+    /// --------
+    /// ```rust
+    /// # use nalgebra_sparse::io::load_coo_from_mm_str;
+    /// # use nalgebra_sparse::io::MMErrorKind;
+    /// let str = r#"
+    /// %%matrixmarket matrix coordinate integer symmetric
+    /// 5 5 2
+    /// 1 1 10
+    /// 2 3 5
+    /// "#;
+    /// let matrix_error = load_coo_from_mm_str::<i32>(str);
+    /// assert_eq!(matrix_error.is_err(), true);
+    /// assert_eq!(*matrix_error.unwrap_err().kind(),MMErrorKind::LowerTriangleError);
+    /// ```
+    LowerTriangleError,
+}
+
+impl MMError {
+    fn from_kind_and_message(error_type: MMErrorKind, message: String) -> Self {
+        Self {
+            error_kind: error_type,
+            message,
+        }
+    }
+
+    /// The operation error kind.
+    #[must_use]
+    pub fn kind(&self) -> &MMErrorKind {
```

Comment:
Maybe return the type directly rather than a reference here, as it gives us a bit more freedom in the future

Reply:
This is because I see that `nalgebra_sparse::ops::serial::OperationError.kind()` and `nalgebra_sparse::SparseFormatError.kind()` both return a reference. So this is why I returned it with a reference. But `std::io::Error` returns the `ErrorKind` not the reference. 

I can fix this, but I think it would be better also change the `nalgebra_sparse::ops::serial::OperationError.kind()` and `nalgebra_sparse::SparseFormatError.kind()`

Reply:
Ah, good point. I'm not sure why I returned the `kind` by reference there...

Let's maybe leave changing `OperationError` and `SparseFormatError` for a separate PR, since this is a breaking change (whereas the matrix market stuff is not).